src/core/agent/services/ipexCommunicationService.ts

```diff
@@ -0,0 +1,266 @@
+    const cred = await this.waitForAcdcToAppear(credentialId);
+    const credentialShortDetails = await this.updateAcdcMetadataRecordCompleted(
+      credentialId,
+      cred
+    );
+    await this.deleteKeriNotificationRecordById(id);
+    this.eventService.emit<AcdcKeriStateChangedEvent>({
+      type: AcdcKeriEventTypes.AcdcKeriStateChanged,
+      payload: {
+        status: CredentialStatus.CONFIRMED,
+        credential: credentialShortDetails,
+      },
+    });
+  }
+
+  private async waitForAcdcToAppear(credentialId: string): Promise<any> {
+    let acdc = await this.getCredentialBySaid(credentialId);
+    let retryTimes = 0;
+    while (!acdc) {
+      if (retryTimes > 120) {
+        throw new Error(IpexCommunicationService.ACDC_NOT_APPEARING);
+      }
+      await new Promise((resolve) => setTimeout(resolve, 500));
+      acdc = (await this.getCredentialBySaid(credentialId)).acdc;
+      retryTimes++;
+    }
+    return acdc;
+  }
+
+  private async getKeriNotificationRecordById(
+    id: string
+  ): Promise<KeriNotification> {
+    const result = await this.basicStorage.findById(id);
+    if (!result) {
+      throw new Error(
+        `${IpexCommunicationService.KERI_NOTIFICATION_NOT_FOUND} ${id}`
+      );
+    }
+    return {
+      id: result.id,
+      createdAt: result.createdAt,
+      a: result.content,
+    };
+  }
+
+  async deleteKeriNotificationRecordById(id: string): Promise<void> {
+    await this.basicStorage.deleteById(id);
+  }
+
+  private async updateAcdcMetadataRecordCompleted(
+    id: string,
+    cred: any
+  ): Promise<CredentialShortDetails> {
+    const metadata =
+      await this.credentialStorage.getCredentialMetadataByCredentialRecordId(
+        id
+      );
+    if (!metadata) {
+      throw new Error(
+        IpexCommunicationService.CREDENTIAL_MISSING_METADATA_ERROR_MSG
+      );
+    }
+
+    metadata.status = CredentialMetadataRecordStatus.CONFIRMED;
+    metadata.credentialType = cred.schema?.title;
+    await this.credentialStorage.updateCredentialMetadata(
+      metadata.id,
+      metadata
+    );
+    return getCredentialShortDetails(metadata);
+  }
+
+  private async createAcdcMetadataRecord(event: any): Promise<void> {
+    await this.saveAcdcMetadataRecord(event.e.acdc.d, event.e.acdc.a.dt);
+  }
+
+  private async saveAcdcMetadataRecord(
+    credentialId: string,
+    dateTime: string
+  ): Promise<void> {
+    const credentialDetails: CredentialShortDetails = {
+      id: `metadata:${credentialId}`,
+      isArchived: false,
+      colors: new ColorGenerator().generateNextColor() as [string, string],
+      credentialType: "",
+      issuanceDate: new Date(dateTime).toISOString(),
+      status: CredentialMetadataRecordStatus.PENDING,
+    };
+    await this.createMetadata({
+      ...credentialDetails,
+      credentialRecordId: credentialId,
+    });
+  }
+
+  async getUnhandledCredentials(): Promise<KeriNotification[]> {
```

Comment:
I don't think this function is used, so it can be removed. And `getKeriCredentialNotifications` too

`getKeriCredentialNotifications` in the credential service is what is being used by the app wrapper